sdk/eventhub/Microsoft.Azure.WebJobs.Extensions.EventHubs/src/Listeners/EventHubListener.PartitionProcessor.cs

```diff
@@ -203,22 +202,24 @@ public async Task ProcessEventsAsync(EventProcessorHostPartition context, IEnume
                     {
                         UpdateCheckpointContext(events, context);
                         await TriggerExecute(events, context, linkedCts.Token).ConfigureAwait(false);
-                        _firstFunctionExecute = false;
                         eventToCheckpoint = events.LastOrDefault();
                     }
 
                     // If total events is less than the batch size, leave them in the stored events list
                     // and wait to send until we receive enough events or total max wait time has passed.
                 }
 
-                // Checkpoint if we processed any events.
+                // Checkpoint if we processed any events and cancellation has not been signaled.
                 // Don't checkpoint if no events. This can reset the sequence counter to 0.
                 // Note: we intentionally checkpoint the batch regardless of function
                 // success/failure. EventHub doesn't support any sort "poison event" model,
                 // so that is the responsibility of the user's function currently. E.g.
                 // the function should have try/catch handling around all event processing
                 // code, and capture/log/persist failed events, since they won't be retried.
-                if (eventToCheckpoint != null)
+                // Don't checkpoint if cancellation has been requested as this can lead to data loss,
+                // since the user may not actually process the event.
+
+                if (eventToCheckpoint != null && !linkedCts.IsCancellationRequested)
```

Comment:
We should check for cancellation of `backgroundCancellationTokenSource` on L279 as well, no?

I think this falls into the same bucket; if the token is signaled while the function code is executing, we can't be sure that the events were processed.   Better to avoid checkpointing and allow the batch to be read again.